### PR TITLE
LibWeb: Let's solve all the reference cycles by moving things to the GC heap

### PR DESCRIPTION
**Pull request by awesomekling**
We have a lot of problems with reference cycles in LibWeb. The main issue is that the web platform is specified assuming everything is garbage collected, and spec objects point at each other left and right. Until now we've been solving this with "one strong + one weak" bidirectional pointers. It's been a continuous uphill battle, and one that we're totally losing -- we're currently leaking the whole world on every page due to numerous ref cycles.

To solve this, let's move all these objects over to the JS GC heap. We do that by removing the JS wrappers in `Web::Bindings` and making the web platform objects inherit `JS::Object` directly instead. Since all the functionality is on the prototype (and we'll keep generating the prototype classes!), this is actually rather straightforward. However, it is a *lot* of work..

### CODE REVIEW

File: Meta/CMake/libweb_generators.cmake [hunk 1]
```diff
@@ -106,23 +106,34 @@ function (generate_js_wrappers target)
 
     function(libweb_js_wrapper class)
         cmake_parse_arguments(PARSE_ARGV 1 LIBWEB_WRAPPER "ITERABLE" "" "")
+        cmake_parse_arguments(PARSE_ARGV 1 LIBWEB_WRAPPER "NO_INSTANCE" "" "")
```
Comment (ADKaster) [inline Meta/CMake/libweb_generators.cmake:109]: Not sure if this first commit will stay at the end of the PR, but NO_INSTANCE should be added to the first call.

`cmake_parse_arguments(PARSE_ARGV 1 LIBWEB_WRAPPER "ITERABLE;NO_INSTANCE" "" "")`

Comment (awesomekling) [reply Meta/CMake/libweb_generators.cmake:109]: Indeed, the idea is for this to be gone by the end of the PR :)

**Comment by circl-lastname** [comment]
Is the "plan" to ever move this out of LibJS? It seems weird for me to use it just for garbage collecting the web engine, feels like a LibGC could be made and used by both LibWeb and LibJS.

**Comment by linusg** [comment]
The Web API objects in LibWeb are regular JS objects, so they naturally already use LibJS's GC mechanism. This only changes that the underlying objects are JS objects themselves instead of being refcounted C++ instances that are wrapped by a separate GC'd JS object forwarding operations to the underlying class.

**Comment by circl-lastname** [comment]
Hm, interesting

File: Userland/Libraries/LibJS/Runtime/FinalizationRegistry.h [hunk 2]
```diff
@@ -41,7 +42,7 @@ class FinalizationRegistry final
 
     virtual void visit_edges(Visitor& visitor) override;
 
-    Handle<Realm> m_realm;
+    NonnullGCPtr<Realm> m_realm;
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/FinalizationRegistry.h:45]: Forgot to visit this though?

Comment (awesomekling) [reply Userland/Libraries/LibJS/Runtime/FinalizationRegistry.h:45]: :catdogalert:

**Review by linusg — COMMENTED** [review]
Woohoo! Made it pas some of the first "make Foo GC-allocated" commits and it seems to get pretty repetitive from there - I'll check out a few more random ones tomorrow.

Typo in `LibWeb: Add Bindings::LegacyPlatformObject base class`: "HTML spec" -> "IDL spec"

~~The LegacyPlatformObject commit seems a bit sloppy, but I think you mentioned it's partially generated code so might just need a bit of cleanup :)
(most of the comments I left there are repetitive, too)~~

**Review by linusg — APPROVED** [review]
As most of the comments are for the generated code and I'm happy to clean these up later, this LGTM!

File: Userland/Libraries/LibWeb/Bindings/MainThreadVM.cpp [hunk 3]
```diff
@@ -307,12 +307,27 @@ JS::VM& main_thread_vm()
...
         auto& custom_data = *verify_cast<WebEngineCustomData>(vm->custom_data());
-        custom_data.root_execution_context = make<JS::ExecutionContext>(vm->heap());
+        custom_data.root_execution_context = MUST(JS::Realm::initialize_host_defined_realm(
+            *vm, [&](JS::Realm& realm) -> JS::GlobalObject* {
+                auto internal_window = HTML::Window::create();
+                custom_data.internal_window_object = JS::make_handle(vm->heap().allocate<Bindings::WindowObject>(realm, realm, internal_window));
+                return custom_data.internal_window_object.cell(); },
+            [](JS::Realm&) -> JS::GlobalObject* {
+                return nullptr;
+            }));
```
[the remark this answers is not in the document]
Comment (awesomekling) [reply Userland/Libraries/LibWeb/Bindings/MainThreadVM.cpp:313]: Yeah that's fair, we can just pass `nullptr` instead.

**Pull request merged**